tests/specs/functional/lemmas-spec.k

```diff
@@ -27,14 +27,16 @@ module LEMMAS-SPEC
  // -----------------------
 
     rule <k> runLemma ( #lookupMemory((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY)    ) => doneLemma ( 33  ) ... </k>
-    // rule <k> runLemma ( #lookupMemory((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY')   ) => doneLemma ( 216 ) ... </k>
-    // rule <k> runLemma ( #lookupMemory((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY'')  ) => doneLemma ( 5   ) ... </k>
-    // rule <k> runLemma ( #lookupMemory((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY''') ) => doneLemma ( 0   ) ... </k>
+    rule <k> runLemma ( #lookupMemory((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY')   ) => doneLemma ( 216 ) ... </k>
+    rule <k> runLemma ( #lookupMemory((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY'')  ) => doneLemma ( 5   ) ... </k>
+    rule <k> runLemma ( #lookupMemory((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY''') ) => doneLemma ( 0   ) ... </k>
+    rule <k> runLemma ( #lookupMemory((KEY |-> 33), KEY') ) => doneLemma ( 0 ) ... </k> requires notBool KEY' in_keys(KEY |-> 33)
 
     rule <k> runLemma ( #lookup((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY)    ) => doneLemma ( 33  ) ... </k>
     rule <k> runLemma ( #lookup((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY')   ) => doneLemma ( 728 ) ... </k>
-    // rule <k> runLemma ( #lookup((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY'')  ) => doneLemma ( 5   ) ... </k>
-    // rule <k> runLemma ( #lookup((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY''') ) => doneLemma ( 0   ) ... </k>
+    rule <k> runLemma ( #lookup((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY'')  ) => doneLemma ( 5   ) ... </k>
+    rule <k> runLemma ( #lookup((KEY |-> 33) (KEY' |-> 728) (KEY'' |-> (pow256 +Int 5)) (KEY''' |-> "hello"), KEY''') ) => doneLemma ( 0   ) ... </k>
+    rule <k> runLemma ( #lookup((KEY |-> 33), KEY') ) => doneLemma ( 0 ) ... </k> requires notBool KEY' in_keys(KEY |-> 33)
```

Comment:
@denis-bogdanas @ehildenb I'm not sure how all these work.  We didn't explicitly say that the keys are distinct.  Is it implicitly assumed by the definedness condition of the Map concatenation?  What if some keys are same?  Then, `runLemma ( #lookup ( #Bottom ) ) => ???`

Reply:
Oh, essentially, the LHS of the spec is split into two terms, depending whether all keys are distinct, or not. That is, we have the following two initial terms:

- runLemma ( #lookup( K1 |-> V1 K2 |-> V2, K1 ) )  /\  K1 =/= K2
- runLemma ( #lookup( #Bottom ) )  /\  K1 == K2

and the second term is silently ignored because the whole term is `#Bottom`.